src/json.rs

```diff
@@ -196,3 +214,148 @@ pub fn parse_json(data: &[u8]) -> Option<JsonValue> {
+        let err = json.u128("foo").unwrap_err();
+        assert_eq!(err, JsonError::ExpectedStringGotNumber);
+
+        let json = parse_json(format!(r#"{{"foo": "{}"}}"#, "bar").as_bytes()).unwrap();
+        let err = json.u128("foo").unwrap_err();
+        assert_eq!(err, JsonError::InvalidU128);
+
+        let json = parse_json(format!(r#"{{"foo": {}}}"#, true).as_bytes()).unwrap();
+        let err = json.u128("foo").unwrap_err();
+        assert_eq!(err, JsonError::InvalidU128);
+
+        let json = parse_json(format!(r#"{{"foo": ["{}"]}}"#, 123).as_bytes()).unwrap();
+        let err = json.u128("foo").unwrap_err();
+        assert_eq!(err, JsonError::InvalidU128);
+
+        let json = JsonValue::Null;
+        let err = json.u128("foo").unwrap_err();
+        assert_eq!(err, JsonError::NotJsonType);
+    }
+
+    #[test]
+    fn test_json_type_bool() {
+        let json = parse_json(format!(r#"{{"foo": {}}}"#, true).as_bytes()).unwrap();
+        let val = json.bool("foo").ok().unwrap();
+        assert_eq!(val, true);
+
+        let json = parse_json(format!(r#"{{"foo": {}}}"#, false).as_bytes()).unwrap();
+        let val = json.bool("foo").ok().unwrap();
+        assert_eq!(val, false);
+
+        let json = parse_json(format!(r#"{{"foo": "{}"}}"#, true).as_bytes()).unwrap();
+        let err = json.bool("foo").unwrap_err();
+        assert_eq!(err, JsonError::InvalidBool);
+
+        let json = parse_json(format!(r#"{{"foo": "{}"}}"#, false).as_bytes()).unwrap();
+        let err = json.bool("foo").unwrap_err();
+        assert_eq!(err, JsonError::InvalidBool);
+
+        let json = parse_json(format!(r#"{{"foo": [{}]}}"#, true).as_bytes()).unwrap();
+        let err = json.bool("foo").unwrap_err();
+        assert_eq!(err, JsonError::InvalidBool);
+
+        let json = parse_json(format!(r#"{{"foo": {}}}"#, 123).as_bytes()).unwrap();
+        let err = json.bool("foo").unwrap_err();
+        assert_eq!(err, JsonError::InvalidBool);
+
+        let json = parse_json(format!(r#"{{"foo": "{}"}}"#, "bar").as_bytes()).unwrap();
+        let err = json.bool("foo").unwrap_err();
+        assert_eq!(err, JsonError::InvalidBool);
+
+        let json = JsonValue::Null;
+        let err = json.bool("foo").unwrap_err();
+        assert_eq!(err, JsonError::NotJsonType);
+    }
+
+    #[test]
+    fn test_json_type_u8() {
+        let json = JsonValue::from(123f64);
+        let val = JsonValue::parse_u8(&json).ok().unwrap();
+        assert_eq!(val, 123);
+
+        let json = JsonValue::from(256f64);
+        let val = JsonValue::parse_u8(&json).ok().unwrap();
+        assert_eq!(val, 255);
+
+        let json = JsonValue::from(260f64);
+        let val = JsonValue::parse_u8(&json).ok().unwrap();
+        assert_eq!(val, 255);
+
+        let json = JsonValue::from(12345678f64);
+        let val = JsonValue::parse_u8(&json).ok().unwrap();
+        assert_eq!(val, 255);
```

Comment:
Probably this is wrong that values over `U8::MAX` are allowed for parsing and return 255 by default.
I think this should be treated as an error, so I will fix this.

Reply:
Fixed out-of-range allowed behaviour for `u8` and `u64` in 973c225.